src/main/java/org/assertj/core/internal/Iterables.java

```diff
@@ -1124,24 +1124,33 @@ public void assertContainsAll(AssertionInfo info, Iterable<?> actual, Iterable<?
   public void assertContainsExactly(AssertionInfo info, Iterable<?> actual, Object[] values) {
     checkIsNotNull(values);
     assertNotNull(info, actual);
-
+    // use actualAsList instead of actual in case actual is a singly-passable iterable
     List<Object> actualAsList = newArrayList(actual);
-    IterableDiff<Object> diff = diff(actualAsList, asList(values), comparisonStrategy);
-    if (!diff.differencesFound()) {
-      // actual and values have the same elements but are they in the same order ?
-      int i = 0;
-      // use actualAsList instead of actual in case actual is a singly-passable iterable
-      for (Object elementFromActual : actualAsList) {
-        if (!areEqual(elementFromActual, values[i])) {
+
+    // length check
```

Comment:
you need to check the size first otherwise you can't be sure that `values[i]` is not going to blow up with an IndexOutOfBoundException.

I think extracting a private method with the common diff + failure code might be better